pageserver/src/tenant/timeline.rs

```diff
@@ -4176,6 +4177,398 @@ impl Timeline {
+
+        while let Some(res) = tasks.join_next().await {
+            match res {
+                Ok(Some(timeline)) => {
+                    tracing::info!(reparented=%timeline.timeline_id, "reparenting done");
+                    reparented.push(timeline.timeline_id);
+                }
+                Ok(None) => {
+                    // lets just ignore this for now. one or all reparented timelines could had
+                    // started deletion, and that is fine.
+                }
+                Err(je) if je.is_cancelled() => unreachable!("not used"),
+                Err(je) if je.is_panic() => {
+                    // ignore; it's better to continue with a single reparenting failing (or even
+                    // all of them) in order to get to the goal state... though, this would be
+                    // exceptional, and we have already done N non-cancellable actions. hard to say
+                    // which requires less manual repair.
+                }
+                Err(je) => tracing::error!("unexpected join error: {je:?}"),
+            }
+        }
+
+        if reparenting_candidates != reparented.len() {
+            tracing::info!("failed to reparent some candidates");
+        }
+
+        rtc.schedule_detaching_from_ancestor_and_wait((ancestor.timeline_id, ancestor_lsn))
+            .await?;
+
+        Ok(reparented)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum DetachFromAncestorError {
+    #[error("no ancestors")]
+    NoAncestor,
+    #[error("too many ancestors")]
+    TooManyAncestors,
+    #[error("shutting down, please retry later")]
+    ShuttingDown,
+    #[error("detached timeline must receive writes before the operation")]
+    DetachedTimelineNeedsWrites,
+    #[error("flushing failed")]
+    FlushAncestor(#[source] anyhow::Error),
+    #[error("layer download failed")]
+    RewrittenDeltaDownloadFailed(#[source] anyhow::Error),
+    #[error("copying LSN prefix locally failed")]
+    CopyDeltaPrefix(#[source] anyhow::Error),
+    #[error("upload rewritten layer")]
+    UploadRewritten(#[source] anyhow::Error),
+
+    #[error("ancestor is already being detached by: {}", .0)]
+    OtherTimelineDetachOngoing(TimelineId),
+
+    #[error("remote copying layer failed")]
+    CopyFailed(#[source] anyhow::Error),
+
+    #[error("unexpected error")]
+    Unexpected(#[source] anyhow::Error),
+}
+
+pub(crate) struct PreparedTimelineDetach {
+    progress_guard: utils::completion::Completion,
+    layers: Vec<Layer>,
+}
+
+#[derive(Debug)]
+pub(crate) struct AncestorDetachOptions {
+    pub(crate) rewrite_concurrency: std::num::NonZeroUsize,
+    pub(crate) copy_concurrency: std::num::NonZeroUsize,
+}
+
+impl Default for AncestorDetachOptions {
+    fn default() -> Self {
+        Self {
+            rewrite_concurrency: std::num::NonZeroUsize::new(2).unwrap(),
+            copy_concurrency: std::num::NonZeroUsize::new(10).unwrap(),
+        }
+    }
```

Comment:
I know it's more pain / less elegant, but please create a TODO to make these PageServerConfig-urable. Not needed as part of this PR, but, I want this option to be controllable via config.
(I think the per-request configurability will not be useful in practice because we don't control the cplane code that will call the API)

Reply:
I can add the TODO, but I agree. Though I hope to find better values by staging usage. 7669836a4

Reply:
Ok, please add it to your TODO and resolve this comment.